rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,1625 @@
+test categories, but the scope is broader than just testing Vector binary. We
+want to test the whole integration.
+
+Ideally we want to test everything: the correctness, performance and
+reliability. Correctness tests are relatively easy, however, it's not yet clear
+how to orchestrate the performance and reliability tests. Measuring performance
+in clusters is quite difficult and requires insight thought to make it right.
+For example, we have to consider and control a lot more variables of the
+environment - like CNI driver, underlying network topology and so on - to
+understand the conditions we're testing. Reliability tests also require
+more careful designing the test environment.
+For this reason, the initial Kubernetes integration only focuses on correctness
+tests. Once we get som experience with correctness test we can expand our test
+suite with tests from other categories.
+
+It is important that we do actually test correctness on all the configurations -
+see this [comment][why_so_much_configurations] as an example. Kubernetes is has
+a lot of LOC, is very complex and properly supporting it is quite a challenge.
+
+The exact design of the tests is an implementation detail, so it's not specified
+in this RFC, but the suggested approach, as a starting point, could be to deploy
+Vector using our documented installation methods, then run some log-generating
+workload and then run assertions on the collected logs.
+
+The things we'd generally want to ensure work properly include (but are not
+limited to):
+
+- basic log collection and parsing
+- log message filtering (both by file paths and by metadata)
+- log events enhancement with metadata
+- partial log events merging
+
+We want the assertions and tests to be cluster-agnostic, so that they work with
+any supplied kubectl config.
+
+#### Existing k8s tests
+
+We already have k8s integration tests implemented in Rust in the Vector repo.
+Currently, they're being run as part of the `cd tests; make tests`. They
+assert that Vector code works properly by deploying Vector plus some test
+log producers and asserting that Vector produced the expected output. This is
+very elegant solution.
+However, these tests are really more like unit tests - in a sense that they
+completely ignore the YAMLs and Helm Charts and
+use their own test configs. While they do a good job in what they're built for -
+we probably shouldn't really consider them integration tests in a broad sense.
+
+It was discussed that we'd want to reuse them as our integration tests, however,
+for the reasons above I don't think it's a good idea. At least as they're now.
+We can decouple the deployment of Vector from the deployment of test containers
+and assertions - then we use just the second half with Vector deployed via YAMLs
+and/or Helm Charts. For now, we should probably leave them as is, maintain them,
+but hold the adoption as integration tests.
```

Comment:
I'm ok deleting these tests. It sounds like you have a good approach planned. We just need to make sure that:

1. Developers can quickly, and easily, run unit tests to ensure they didn't break the integration.
2. Run local integration tests against minikube, or something light-weight, to further validate integration.
3. And then once all of these pass, command full integration tests against select real-world clusters as a final step.